### PR TITLE
Some Utility Umbrella fixes

### PR DESCRIPTION
**Pull request by urkerab**
- Cherrim should revert if given a Utility Umbrella in Sun.
- Castform shouldn't keep transforming if it holds a Utility Umbrella.
- Leaf guard Utility Umbrella check works but is just inconsistent placement.
- Sun/Rain affected recovery moves should heal 50% with Utility Umbrella.
- Solar moves power should halve in sand or hail even with Utility Umbrella.
- Weather Ball should double in power if Magic Room blocks Utility Umbrella.

### CODE REVIEW

**Comment by urkerab** [comment]
This started getting repetitive and I was wondering whether it was worth having a `pokemon.effectiveWeather()` function which checks for the umbrella and if so ignores sun and rain.

File: data/moves.js [hunk 1]
```diff
@@ -17576,8 +17572,7 @@ let BattleMovedex = {
 			return null;
 		},
 		onBasePower(basePower, pokemon, target) {
-			if (pokemon.hasItem('utilityumbrella')) return;
-			if (this.field.isWeather(['raindance', 'primordialsea', 'sandstorm', 'hail'])) {
+			if (this.field.isWeather(pokemon.hasItem('utilityumbrella') ? ['sandstorm', 'hail'] : ['raindance', 'primordialsea', 'sandstorm', 'hail'])) {
```
Comment (KrisXV) [inline data/moves.js]: I don't really like how long this line is, maybe changing it to a variable? it functions the same either way but using a separate function has more readability

File: data/moves.js [hunk 2]
```diff
@@ -20904,12 +20896,12 @@ let BattleMovedex = {
...
 			case 'sunnyday':
 			case 'desolateland':
-				if (pokemon.item === "utilityumbrella") break;
+				if (pokemon.hasItem('utilityumbrella')) break;
 				move.basePower *= 2;
 				break;
 			case 'raindance':
 			case 'primordialsea':
-				if (pokemon.item === "utilityumbrella") break;
+				if (pokemon.hasItem('utilityumbrella')) break;
```
Comment (KrisXV) [inline data/moves.js]: i swear i fixed these i dont know what happened when i rebased last time

**Comment by Zarel** [comment]
I definitely approve of `pokemon.hasWeather`

**Comment by Zarel** [comment]
Although I would name it `pokemon.hasWeather`

**Comment by urkerab** [comment]
Still using `effectiveWeather` for now since some of the `switch` statements are tricky to implement using `hasWeather`.

**Comment by urkerab** [comment]
Ugh, accidentally added the comment with spaces instead of tabs...

**Comment by Zarel** [comment]
Looks good. Thanks!

**Pull request merged**